text/0000-dynamic-import-hints.md

```diff
@@ -0,0 +1,246 @@
+* Hint key and value must be statically verifiable i.e. no curly braces to implicitly refer to a variable and such
+* Multiple hints will be allowed but with restrictions
+* Hints do not influence the outcome of the compilation phase. It only produces additional metadata for the application framework to consume.
+
+## Detailed Design
+
+### Grammar
+There will be grammar restriction in place for hints so that the compiler can parse the LWC modules, gather the metadata and provide that as a result of the compilation.
+
+#### Hint Syntax
+A hint is expressed as a comment inside a dynamic import expression. The hint syntax is either a line comment or a block comment with a *key* as a resource and a *value* as a resolved resource value. The key and value will be separated by a colon(`:`) and delimited using double quotes(`"`).
+
+```js
+/* "<key>": "<value>" */
+```
+Hint comment e.g.:
+```js
+/* "@salesforce/client/formFactor": "SMALL" */
+/* "MY_CONDITION": "VALID" */
+// "MY_OTHER_CONDITION": "NOT_VALID"
+```
+
+Dynamic import with a hint e.g.:
+```js
+import formFactor from '@salesforce/client/formFactor';
+
+export async function loadFoo() {
+    if (formFactor === 'SMALL') {
+        return import("lightning-foo-mobile" /* "@salesforce/client/formFactor": "SMALL" */);
+    } else {
+        return import("lightning-foo-desktop" /* "@salesforce/client/formFactor": "LARGE" */);
+    }
+}
+```
+
+#### Static Semantics
+* Both key and value have to be string literal
+* Is case sensitive
+* Key and value have to be delimited with a double quote to be explicit about the literal value
+* Spaces before and after the colon(`:`) is optional
+* Hint phrase can have leading and trailing spaces
+
+##### Hint Key:
+* First character is an alphabet, can be uppercase or lower case. Can also start with `@` character
+* Cannot contain any special characters except for underscore(`_`), hyphen(`-`), dot(`.`) and forward slash(`/`)
+* Begins and ends with a double quote(`"`)
+* Cannot contain spaces, cannot be delimited by a single quote(`'`)
+
+##### Hint value:
+* Can contain alphanumberic characters
+* Can contain space, underscore(`_`), hyphen(`-`), at sign(`@`), forward slash(`/`)
+* Cannot combine multiple values using operators(i.e `"KEY": "LARGE"&"SMALL"` or `"KEY": "LARGE && SMALL"` will not qualify as a valid hint)
+
+##### Multiple Hints:
+* Each hint must be in a separate block comment or line comment
+* Multiple hints cannot be combined in a single comment using operator
+* Each hint will be specified in a separate comment block
+* They cannot be combined using operators like AND or &&.
+
+#### Semantics
+The [_ModuleSpecifier_](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-imports) of the import statement should be a string literal. The compiler cannot analyze data about a module name that will be discovered or learned at runtime. For example, in the following case no hint metadata will be gathered:
```

Comment:
Isn't this already defined here? Isn't that out of scope for the hints proposal?

https://github.com/salesforce/lwc-rfcs/blob/8e33931ddec7f55216489ec636a4ad98b77a5c11/text/0110-dynamic-components.md#L296

Reply:
The [dynamic imports implementation](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/babel-plugin-component/src/dynamic-imports.js#L55-L68) allows for this rule to be relaxed.